datastream-common/src/main/java/com/linkedin/datastream/connectors/CommonConnectorMetrics.java

```diff
@@ -186,23 +187,30 @@ protected void deregisterAggregates() {
     static final String NUM_PARTITIONS = "numPartitions";
 
     // Per consumer metrics
-    final AtomicLong _numStuckPartitions = new AtomicLong(0);
-    final AtomicLong _numPartitions = new AtomicLong(0);
     final Meter _rebalanceRate;
 
     // Aggregated metrics
     final Meter _aggregatedRebalanceRate;
 
+    final String _fullMetricsKey;
+
     // Map from connector class name to its stuck and total partition counter
     // This is needed for Gauge metrics which need long-typed suppliers.
+    static final Map<String, AtomicLong> NUM_STUCK_PARTITIONS_PER_METRIC_KEY = new ConcurrentHashMap<>();
+    static final Map<String, AtomicLong> NUM_PARTITIONS_PER_METRIC_KEY = new ConcurrentHashMap<>();
     static final Map<String, AtomicLong> AGGREGATED_NUM_STUCK_PARTITIONS = new ConcurrentHashMap<>();
     static final Map<String, AtomicLong> AGGREGATED_NUM_PARTITIONS = new ConcurrentHashMap<>();
+    private long _numStuckPartitions = 0;
+    private long _numPartitions = 0;
```

Comment:
Once concern I have here is what if two threads update these metrics at the same time? Since these aren't atomic anymore, will that be something to worry about? Same for the numTopics variable for the numTopics metric.

Reply:
changed it to AtomicLong.